crates/aptos-in-memory-cache/src/lib.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use std::hash::Hash;
+
+pub mod caches;
+
+/// A trait for a cache that can be used to store key-value pairs.
+pub trait Cache<K, V>: Send + Sync
+where
+    K: Eq + Hash + Clone + Send + Sync,
+    V: Clone + Send + Sync,
+{
+    /// Get the value for a given key. Return [`None`] if the key is not in the cache.
+    fn get(&self, key: &K) -> Option<V>;
+
+    /// Inserts a given key-value pair in cache. Panics if the insert fails.
+    fn insert(&self, key: K, value: V);
+
+    /// Returns the total size of the cache.
+    fn total_size(&self) -> u64;
+}
+
+pub trait Ordered<K>: Send + Sync
+where
+    K: Clone + Send + Sync,
+{
+    /// Returns the first key in the cache. Returns [`None`] if the cache is empty.
+    fn first_key(&self) -> Option<K>;
+
+    /// Returns the last key in the cache. Returns [`None`] if the cache is empty.
+    fn last_key(&self) -> Option<K>;
+}
+
+// TODO: Find better solution for guaranteeing order
+pub trait Incrementable<V>: Send + Sync
```

Comment:
what's the usage case of this trait binding with `V`?

Reply:
`V` represents the value and is there to provide context to the `next()` function in case we also need the value of the key to increment it. 

For example, the Event Stream uses `(transaction_version, event_index)` as the key for each event. The value provides a `num_events` for us determine that whether to increment `transaction_version` or `event_index` to get the next key.